src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java

```diff
@@ -106,72 +266,186 @@ public void run() {
                     } else if (command instanceof PercentType) {
-                        level = ((PercentType) command).intValue();
+                        PercentType brightness = (PercentType) command;
+                        changeBrightness(brightness);
                     } else if (command instanceof OnOffType) {
-                        level = (OnOffType) command == OnOffType.ON ? 100 : 0;
-                    } else {
-                        return;
+                        OnOffType onoff = (OnOffType) command;
+                        changeOnOff(onoff);
                     }
-
-                    clusterLevelControl.moveToLevelWithOnOffCommand((int) (level * 254.0 / 100.0 + 0.5), 10).get();
-                } catch (InterruptedException e) {
-                    e.printStackTrace();
-                } catch (ExecutionException e) {
-                    e.printStackTrace();
+                } catch ( /* InterruptedException | ExecutionException | */ Exception | Error e) {
+                    logger.warn("{}: Exception processing command", endpoint.getIeeeAddress(), e);
                 }
             }
         };
     }
 
     @Override
     public Channel getChannel(ThingUID thingUID, ZigBeeEndpoint endpoint) {
-        if (endpoint.getInputCluster(ZclColorControlCluster.CLUSTER_ID) == null
-                || endpoint.getInputCluster(ZclLevelControlCluster.CLUSTER_ID) == null) {
+        clusterColorControl = (ZclColorControlCluster) endpoint.getInputCluster(ZclColorControlCluster.CLUSTER_ID);
+
+        if (clusterColorControl == null) {
             return null;
         }
+
+        try {
+            if (!clusterColorControl.discoverAttributes(false).get()) {
+                logger.warn(
+                        "{}: Cannot determine whether device supports RGB color. Assuming it does by now (checking again later)",
+                        endpoint.getIeeeAddress());
+            } else if (!clusterColorControl.getSupportedAttributes().contains(ZclColorControlCluster.ATTR_CURRENTHUE)
+                    && !clusterColorControl.getSupportedAttributes().contains(ZclColorControlCluster.ATTR_CURRENTX)) {
+                return null;
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            logger.warn(
+                    "{}: Exception checking whether device supports RGB color. Assuming it does by now (checking again later)",
+                    endpoint.getIeeeAddress(), e);
+        }
+
         return createChannel(thingUID, endpoint, ZigBeeBindingConstants.CHANNEL_COLOR_COLOR,
                 ZigBeeBindingConstants.ITEM_TYPE_COLOR, "Color");
     }
 
+    private void updateOnOff(OnOffType onoff) {
+        updateBrightness(lastBrightness);
+    }
+
+    private void updateBrightness(PercentType brightness) {
+        HSBType oldHSB = currentHSB;
+        HSBType newHSB = new HSBType(oldHSB.getHue(), oldHSB.getSaturation(), brightness);
+        currentHSB = newHSB;
```

Comment:
I don't think we need the extra step here - I think you can just use this? -:
```
currentHSB - new HSBType(...).
```

Reply:
If we remove oldHSB we will need synchronization in all the code that might change currentHSB, as currentHSB might change between getHue() and getSaturation(), leading to unconsistent / unpredicatable resulting color.

As newHSB is used also later, if we use currentHSB intead woukld also lead to inconsistencies between the received parameter (ie brightnes here), and currentHSB if modified, requiring again synchronization in all the code modifying current HSB.

I would prefer to keep the extra step and both temp variables than synchronizing all currentHSB changes just for this. What do you think ?

Reply:
Ok, I didn't consider the thread synch issues here so I guess it's ok to leave this. Maybe add a comment to remind anyone reviewing later.